### ISSUE TITLE
[Download-Artifacts] CDN Issues?

### ISSUE THREAD

**Issue by ChrisBlankDe** [microsoft/navcontainerhelper#2934]
**Describe the issue**
since a few days we get problems with the download of artifacts over the CDN. The direct access to the blob works very well and fast.

unfortunately/ fortunately the problem does not always occur.
we could observe the problem in different infrastructures/networks.

**Scripts used to create container and cause the issue**
```
$ArtifactUrl = Get-BCArtifactUrl -select Weekly
Download-Artifacts -artifactUrl $ArtifactUrl
```

**Full output of scripts**
```
BcContainerHelper version 4.0.14
Setting defaultNewContainerParameters = @{shortcuts=CommonDesktopFolder}
BC.HelperFunctions emits usage statistics telemetry to Microsoft

Downloading application artifact /sandbox/21.5.53619.54140/at
https://bcartifacts.azureedge.net/sandbox/21.5.53619.54140/at
Downloading C:\Users\AzDevOps\AppData\Local\Temp\048ca889-0cd5-4800-a16a-d3a801741438.zip
Using WebClient
Unpacking application artifact to tmp folder using 7zip

ERRORS:
Unexpected end of archive

ERROR: Data Error : Applications.AT\Base Application.Source.zip
Error unpacking artifact downloaded from CDN, retrying download from direct download URL
https://bcartifacts.blob.core.windows.net/sandbox/21.5.53619.54140/at
Downloading C:\Users\AzDevOps\AppData\Local\Temp\89ce7fdd-6697-477c-8965-e9410036b38a.zip
Using WebClient
Unpacking application artifact to tmp folder using Expand-Archive
```

**Screenshots**

**Additional context**
- does it happen all the time? No
- did it use to work? Yes

**Comment by DanielGoehler** [comment]
@freddydk +1 
We have seen this error message "Unexpected end of archive" or "Could not download" a lot in the last few days.

```
BcContainerHelper is version 4.0.14
BcContainerHelper is running as administrator
HyperV is Enabled
UsePsSession is True
Host is Microsoft Windows Server 2019 Standard - 10.0.17763.4010
Docker Client Version is 19.03.5
Docker Server Version is 19.03.5
Removing Desktop shortcuts
loading application artifact /sandbox/21.5.53619.54140/de
https://bcartifacts.azureedge.net/sandbox/21.5.53619.54140/de
Downloading C:\Windows\TEMP\44179ce0-9583-4a8c-8773-7be8e58ea830.zip
Using WebClient
Could not download from https://bcartifacts.azureedge.net/..., retrying from https://bcartifacts.blob.core.windows.net/ in 3 seconds...
Using WebClient
Download-Artifacts Telemetry Correlation Id: 05dad3f9-70c1-4074-b176-8084095f2522
New-NavContainer Telemetry Correlation Id: f6dafd4c-fbec-43c6-af56-0cdb55617b95
```

**Comment by AmandaTest** [comment]
I'm having the same issue with US artifacts.  My error looks like this: 
<img width="496" alt="image" src="https://user-images.githubusercontent.com/28625572/226442316-105c830e-ac27-4829-918e-2f958306b88c.png">

**Comment by freddydk** [comment]
If you use:

Get-BcArtifactUrl -storageAccount 'bcartifacts.blob.core.windows.net' -select weekly -country us

Then it will never try the CDN, but will go directly for the storageaccount.
(same with bcinsider and other storageaccounts)

**Comment by mjegob** [comment]
> It would help if you would specify which artifacts you are trying to download (and how)

Sure, sorry forgot to write it. 
I execute the folowing code
```pwsh
$artifactUrl = Get-BCArtifactUrl -type Sandbox -country de -select Latest -storageAccount BCPublicPreview
 $containerCreationParameter = @{
            artifactUrl               = $artifactURL
            containerName             = $Name
            licenseFile               = $LicenseFile
            auth                      = 'NavUserPassword'
            Credential                = $Credential
            shortcuts                 = 'CommonDesktop'
            additionalParameters      = $EnvironmentVariables
            restart                   = 'always'
            accept_eula               = $true
            accept_outdated           = $true
            alwaysPull                = $false
            assignPremiumPlan         = $true
            doNotUseRuntimePackages   = $true
            includePerformanceToolkit = $false
            includeTestToolKit        = $true
            includeTestLibrariesOnly  = $true
            useBestContainerOS        = $true
            useSSL                    = $true
            updateHosts               = $true
            useTraefik                = $useTraefik
            PublicDnsName             = $PublicDnsName
            multitenant               = $true
            isolation                    = 'process'
            myscripts                    = $Scripts
        }

        try {
            New-BcContainer @containerCreationParameter -ErrorAction Stop
        } catch {
            Write-Host "Couldn't create Container"
            Write-Host $_
            return $_
        }
```

My Ouput (This try downloaded the application artifacts directly) 
```log
BcContainerHelper version 4.0.14
BC.HelperFunctions emits usage statistics telemetry to Microsoft
BcContainerHelper is version 4.0.14
BcContainerHelper is running as administrator
HyperV is Disabled
UsePsSession is True
Host is Microsoft Windows Server 2022 Datacenter - 10.0.20348.1607
Docker Client Version is 23.0.1
Docker Server Version is 23.0.1
Removing Desktop shortcuts
Downloading application artifact /sandbox/22.0.54157.54529/de
https://bcpublicpreview.azureedge.net/sandbox/22.0.54157.54529/de
Downloading C:\Users\mje\AppData\Local\Temp\4183fa9b-846d-4fd5-9c03-ef4af0603cbe.zip
Using WebClient
Unpacking application artifact to tmp folder using 7zip
Downloading platform artifact /sandbox/22.0.54157.54529/platform
https://bcpublicpreview.azureedge.net/sandbox/22.0.54157.54529/platform
Downloading C:\Users\mje\AppData\Local\Temp\bfd4a7f6-cdd6-4ad0-bfe4-dbdb18de8533.zip
Using WebClient
```
Then it stucks here and i stopped it after an hour. 



> If you use:
> 
> Get-BcArtifactUrl -storageAccount 'bcartifacts.blob.core.windows.net' -select weekly -country us
> 
> Then it will never try the CDN, but will go directly for the storageaccount. (same with bcinsider and other storageaccounts)

Oh good to know, i'll try this. 😊
Thanks

**Comment by freddydk** [comment]
I will add a setting to BcContainerHelper which you can set to avoid the use of CDN - meaning that it will always go for direct blob storage.
In most cases, this will be slower - but if there are issues with the CDN - it will use that.
When a machine has problems with CDN - then I will set this setting for the session to avoid future usages of CDN.
After a restart or a PS restart - it will retry (unless set in settings file)

Make sense?

**Comment by freddydk** [comment]
Shipped in 5.0.2
